### PR TITLE
fix: allow invoking with array arguments and add some consistency checks

### PR DESCRIPTION
**Pull request by kobigurk**

### CODE REVIEW

File: src/starkware/starknet/cli/starknet_cli.py [hunk 1]
```diff
@@ -106,12 +106,30 @@ async def invoke_or_call(args, command_args, call: bool):
         raise ValueError('Invalid address format.')
     for abi_entry in abi:
         if abi_entry['type'] == 'function' and abi_entry['name'] == args.function:
+            previous_input = None
```
Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:109]:
```suggestion
            previous_felt_input = None
```
It is enough to keep the last felt input :)

File: src/starkware/starknet/cli/starknet_cli.py [hunk 2]
```diff
@@ -106,12 +106,30 @@ async def invoke_or_call(args, command_args, call: bool):
...
+            current_inputs_ptr = 0
+            for input_desc in abi_entry['inputs']:
+                assert current_inputs_ptr < len(args.inputs), \
+                        f'Expected at least {current_inputs_ptr + 1}, got {len(args.inputs)}'
```
Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:113]:
```suggestion
                        f'Expected at least {current_inputs_ptr + 1} inputs, got {len(args.inputs)}.'
```

File: src/starkware/starknet/cli/starknet_cli.py [hunk 3]
```diff
@@ -106,12 +106,30 @@ async def invoke_or_call(args, command_args, call: bool):
...
+                current_input_val = args.inputs[current_inputs_ptr]
```
Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:114]:
```
                assert current_inputs_ptr < len(args.inputs), \
                        f'Expected at least {current_inputs_ptr + 1}, got {len(args.inputs)}'
                current_input_val = args.inputs[current_inputs_ptr]
```
The current input value is only interesting when the argument is of type `felt`, thus this part should be moved to the next condition clause. In addition, the assertion may not hold when the argument is of type `felt*` and it is OK (e.g., when the length of the array is 0 and hence there is no need for a `felt*` argument after it)

Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:114]:
```suggestion
                previous_felt_input = args.inputs[current_inputs_ptr]
```

File: src/starkware/starknet/cli/starknet_cli.py [hunk 4]
```diff
@@ -106,12 +106,30 @@ async def invoke_or_call(args, command_args, call: bool):
...
+                if input_desc['type'] == 'felt':
+                    current_inputs_ptr += 1
+                elif input_desc['type'] == 'felt*':
+                    assert current_inputs_ptr > 0, 'First input type cannot be felt*.'
+                    assert previous_input['desc']['type'] == 'felt', \
+                        'Current input is felt*, and previous input should have been ' \
+                        f'felt, but it was {previous_input["desc"]["type"]}'
```
Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:121]:
```suggestion
                    assert previous_felt_input is not None, (
                        f"The array argument '{input_desc['name']}' of type felt* must be preceded "
                        'by a length argument of type felt.')
```

File: src/starkware/starknet/cli/starknet_cli.py [hunk 5]
```diff
@@ -106,12 +106,30 @@ async def invoke_or_call(args, command_args, call: bool):
...
+
+                    current_inputs_ptr += previous_input['val']
```
Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:123]:
```suggestion
                    current_inputs_ptr += previous_felt_input
```

**Review by Adi-Yakovian — CHANGES_REQUESTED** [review]

File: src/starkware/starknet/cli/starknet_cli.py (hunk 1 above)
Comment (kobigurk) [reply src/starkware/starknet/cli/starknet_cli.py:109]: Good point :)

File: src/starkware/starknet/cli/starknet_cli.py (hunk 3 above)
Comment (kobigurk) [reply src/starkware/starknet/cli/starknet_cli.py:114]: That's very true. It's also incorrect when the invocation doesn't have any arguments. Let me think.

Comment (kobigurk) [reply src/starkware/starknet/cli/starknet_cli.py:114]: Nevermind, it's correct in my case since we won't enter the loop. Thinking about yours.

Comment (Adi-Yakovian) [reply src/starkware/starknet/cli/starknet_cli.py]: I agree :)

File: src/starkware/starknet/cli/starknet_cli.py [hunk 6]
```diff
@@ -106,12 +106,28 @@ async def invoke_or_call(args, command_args, call: bool):
...
+                    assert current_inputs_ptr < len(args.inputs), \
+                        f'Expected at least {current_inputs_ptr + 1} inputs, got {len(args.inputs)}'
+                    previous_felt_input = args.inputs[current_inputs_ptr]
+                    current_inputs_ptr += 1
+                elif input_desc['type'] == 'felt*':
+                    assert previous_felt_input is not None, \
+                        f'The array argument {input_desc["name"]} of type felt* must be preceded ' \
+                        'by a length argument of type felt.'
+
+                    current_inputs_ptr += previous_felt_input
```
Comment (Adi-Yakovian) [inline src/starkware/starknet/cli/starknet_cli.py:122]: I forgot to add it, but the following assignment 
```
previous_felt_input = None
```
should be added to this (`felt*`) case.

Comment (kobigurk) [reply src/starkware/starknet/cli/starknet_cli.py:122]: Oops, right! Thanks!

**Review by Adi-Yakovian — APPROVED** [review]

**Pull request merged**